daemon/graphdriver/copy/copy.go

```diff
@@ -117,22 +202,80 @@ type dirMtimeInfo struct {
 	stat    *syscall.Stat_t
 }
 
+type dstFilePathWithLock struct {
+	lock sync.Mutex
+	once int
+	path string
+}
+
+func newDstFilePathWithLock(path string) *dstFilePathWithLock {
+	ret := &dstFilePathWithLock{
+		path: path,
+	}
+	ret.lock.Lock()
+	return ret
+}
+
+// Not concurrency safe, but done should only be called either if you instantiate
+// the object
+func (d *dstFilePathWithLock) done() {
+	if d.once == 0 {
+		d.lock.Unlock()
+		d.once++
+	}
+}
+
+func (d *dstFilePathWithLock) waitForCopy() {
+	d.lock.Lock()
+	d.lock.Unlock()
+}
+
+type copiedFilesMap struct {
+	copiedFilesMap sync.Map
+}
+
+func (c *copiedFilesMap) getCopiedFileOrNew(id fileID, path string) (*dstFilePathWithLock, bool) {
+	tmpNew := newDstFilePathWithLock(path)
+	val, loaded := c.copiedFilesMap.LoadOrStore(id, tmpNew)
+	return val.(*dstFilePathWithLock), loaded
+}
+
 // DirCopy copies or hardlinks the contents of one directory to another,
 // properly handling xattrs, and soft links
 //
 // Copying xattrs can be opted out of by passing false for copyXattrs.
 func DirCopy(srcDir, dstDir string, copyMode Mode, copyXattrs bool) error {
-	copyWithFileRange := true
-	copyWithFileClone := true
 
 	// This is a map of source file inodes to dst file paths
-	copiedFiles := make(map[fileID]string)
+	copiedFiles := &copiedFilesMap{}
 
 	dirsToSetMtimes := list.New()
+	concurrentFileCopiers := make([]chan fileCopyInfo, concurrentFileCopies)
+
+	// This is a wait-group + cancellation pipeline
+	ctx, cancel := context.WithCancel(context.Background())
+	errGroup, childCtx := errgroup.WithContext(ctx)
+	// Do not return until all of the subroutines are completed
+	defer errGroup.Wait()
+	// defer cancel, so if early return occurs, errGroup is cancelled before return
+	defer cancel()
+
+	for i := range concurrentFileCopiers {
+		concurrentFileCopyChan := make(chan fileCopyInfo, fileCopyChannelDepth)
```

Comment:
This is a pretty massive allocation, have you done any benchmarks which show the speed and allocations? I would be curious to see that and how it changes at different depths.

Reply:
These numbers came about (8 concurrent and 10000 slots in the channel) from empirical testing on the big image with lots of nested files. 8 makes sense, as we were testing on an XFS filesystem, and we found 8 was the number that it hit lock contention at.

Testing with 1000, 5000, 10000, and 15000. -- 15K showed no benefit over 10K. On the other hand 10K showed a benefit over 5K. My *guess* as to why this behaviour was showing up is because as soon as these channels block, we start to pause before creating all the directories. Then you start to get contention between the two.

Reply:
My understanding is that this is allocating 8 * 10000 * sizeof(fileCopyInfo). Sure the memory trade off is likely worth a 2x performance gain. Having such deep channels depths is normally an anti-pattern in Go, it implies there is a need to unblock the sender and variability in the time it takes the receiver to process an item. If the sender is just faster, then the usefulness of this depth depends on the number of elements, meaning for many copies will allocate way more than necessary. What I really want to understand is why the sender needs to be unblocked if the receiver is not ready and why is that having such a significant impact on performance.

Reply:
I'd ask Dave Chinner why some file metadata operations are faster than others. :P. As far as I can tell, it has to do with metadata contention.